### PR TITLE
[experiment] CogVideoX 🤝🏼 FreeNoise

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
# What does this PR do?

Attempt to make FreeNoise work with CogVideoX to enable longer video generation, prompt interpolation (something like Luma AI Keyframes but without image condition control since available Cog models are txt2vid), etc.

<details>
<summary> Code </summary>

```python
import torch
from diffusers import CogVideoXPipeline, CogVideoXDPMScheduler
from diffusers.utils import export_to_video

# Models: "THUDM/CogVideoX-2b" or "THUDM/CogVideoX-5b"
model_id = "THUDM/CogVideoX-2b"

pipe = CogVideoXPipeline.from_pretrained(model_id, torch_dtype=torch.float16)
pipe.scheduler = CogVideoXDPMScheduler.from_config(pipe.scheduler.config)
pipe.to("cuda")

pipe.enable_free_noise()

prompt = (
    "A panda, dressed in a small, red jacket and a tiny hat, sits on a wooden stool in a serene bamboo forest. "
    "The panda's fluffy paws strum a miniature acoustic guitar, producing soft, melodic tunes. Nearby, a few other "
    "pandas gather, watching curiously and some clapping in rhythm. Sunlight filters through the tall bamboo, "
    "casting a gentle glow on the scene. The panda's face is expressive, showing concentration and joy as it plays. "
    "The background includes a small, flowing stream and vibrant green foliage, enhancing the peaceful and magical "
    "atmosphere of this unique musical performance."
)

video = pipe(
    prompt=prompt,
    num_frames=81,
    guidance_scale=6,
    use_dynamic_cfg=True,
    num_inference_steps=50,
).frames[0]

filename = "output.mp4"
export_to_video(video, filename, fps=8)

```
</details>

Here are some preliminary txt2vid results:

<table align=center>
<tr>
  <td align=center><video src="https://github.com/user-attachments/assets/349db656-3bff-4232-aae1-397784f3d344"> Your browser does not support the video tag. </video></td>
  <td align=center><video src="https://github.com/user-attachments/assets/da8cda5b-2b5f-492a-b960-c46077c5c5e5"> Your browser does not support the video tag. </video></td>
</tr>
<tr>
  <td align=center><video src="https://github.com/user-attachments/assets/18129d05-1ecc-4d22-b954-87742af90a7d"> Your browser does not support the video tag. </video></td>
  <td align=center><video src="https://github.com/user-attachments/assets/1899aa86-c872-4399-a9b3-e7805e8e74d5"> Your browser does not support the video tag. </video></td>
</tr>
</table>

I also tried the context windows method with temporal tiling on the transformer in the pipeline but got poorer results compared to FreeNoise. I don't think it's worth considering supporting as we don't support it for AnimateDiff as well. FreeNoise, however, performs better although there is some visible repetition going on for the same prompt. This is most likely due to `noise_type` in above generations being `"shuffle_context"`. I'm expecting improvements when using better initial noise generation strategies.

`context_length` as 13 and `context_stride` as 4 seem to produce the best results on a variety of prompts. This corresponds to 49 and 16 pixel-space frames respectively.

TODO:
- [ ] Add support for prompt interpolation (tricky)
- [ ] Add support for Cog-5b (tricky too, and might require some refactoring of the last normalization layer)
- [ ] Tests
- [ ] Docs

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@DN6 @yiyixuxu @asomoza 

@zRzRzRzRzRzRzR @wenyihong because changes involve CogVideoX

@arthur-qiu @tin2tin for visibility

### CODE REVIEW

**Comment by tin2tin** [comment]
Happy to see work being done on this:
Tried to run the code on a 4090. It is successfully produced a clip. 
- First step takes 8-10 minutes, and needs 20 GB VRAM.
- Second step doesn't have a progress-bar and needs 36 GB VRAM. 
- Even though the process has finished, the 36 GB are still occupying the VRAM. 

https://github.com/user-attachments/assets/ec214add-5942-4395-ad23-a89c12723893

**Comment by TemporalLabsLLC-SOL** [comment]
> Happy to see work being done on this. Tried to run the code on a 4090. It successfully produced a clip.
> 
> * First step takes 8-10 minutes, and needs 20 GB VRAM.
> * Second step doesn't have a progress-bar and needs 36 GB VRAM.
> * Even though the process has finished, the 36 GB are still occupying the (V)RAM.
> 
>  output.mp4

That's not much different than the scene transition I get with the comfyui tiling extend work around tbh.

I'm hoping to find a better way using the Temporal Prompt Engine and Video Gen script during the recombine phase. 

Feel free to check out my repo if you want to help.

**Comment by subin-kim-cv** [comment]
Thank you for your hard work!

I ran the code to generate videos four times longer than the original length (with a frame length of more than 200). While the system works well for generating videos up to twice the original length, I’ve encountered severe artifacts in the outputs when generating longer videos using CogVideoX.

Do you have any insights into why this might happen or suggestions on mitigating these artifacts?

**Comment by a-r-r-o-w** [comment]
Hi, thanks for trying this out! FreeNoise usually works great for models that have `spatial layer - temporal layer - spatial layer - ...`. This is the case with models like AnimateDiff, and such an approach works really well. Trying to apply something similar for CogVideoX, which does not have a dedicated temporal layer, is already a bit of a stretch and I think FreeNoise-like approach is not ideal for extending the video length of tokenization-based transformer models.

If you know about the context sliding window approach generally used in the community for extending videos, that will probably be more useful here and people have reported mixed success with it. But, to improve things here, I think you will have to try a different approach for how the latents are created and averaged. I usually get poor results with random latent noise across full tensor (frame-wise averaging tends to mess with the spatial tokens leading to many artifacts), and a lot of repitition in video is visible when using `repeat_context` or `shuffle_context` to create initial latent. I think it is worth exploring better ways of extending videos that build upon FreeNoise. Unfortunately, I haven't found the time to continue working on this and explore ideas so do not have any good insights at the moment.

If you are interested on working out training-free approaches for extending videos in DiTs, I think we could spend some time together on trying things out. Would you be interested? If so, I would be happy to invite you to our internal Slack with the rest of Diffusers team who we can pool our ideas together with (cc @sayakpaul)

**Comment by subin-kim-cv** [comment]
Thank you for the quick response!

I’m currently working on training-free approaches with new ideas for
extending videos in Diffusion models, including U-Net-based and DiTs. This
work is expected to be published around late November.

I am strongly interested in the project of extending videos with Diffusion
models, and it would be great if you could invite me to your channel.
While I’m a bit busy with the CVPR deadline until November,
I’ll be more available after that and would love to contribute.

Best regards,
Subin
